### PR TITLE
BMV2 Integration

### PR DESCRIPTION
**Pull request by JLin-Li**
Adding SimBricks integration for P4 reference behavior model BMV2. Functionality verified using qemu_udp_single experiment.

To-dos:
  - Add synchronization support
  - Add p4c support to generate simple switch from p4 source code

### CODE REVIEW

**Review by marvin71 — COMMENTED** [review]
In general this looks good to me, but I haven't tested it. I have only one minor comment regarding the clean target.

**Review by jonas-kaufmann — CHANGES_REQUESTED** [review]
Nice, thanks for the PR! Overall looks good to me. Can you squash the commits though using `git rebase`?  If possible we want to have a clean history. Concretely, I'd propose to have one commit that adds the submodule, everything necessary to build bmv2 and the integration in `simulators.py`. In a second commit you could then modify the experiment.

File: sims/external/rules.mk [hunk 1]
```diff
@@ -100,6 +102,19 @@ femu-clean:
...
+	CPPFLAGS=-I$(abspath $(lib_dir)) ./configure && \
+	make -j
+	touch $@
+
+bmv2-clean:
+	-cd $(EXTERNAL_SIMS_DIR)bmv2 && $(MAKE) clean
+	rm -f $(EXTERNAL_SIMS_DIR)bmv2/ready
+
 DISTCLEAN := $(d)gem5 $(d)qemu $(d)ns-3 $(d)femu
 EXTERNAL_CLEAN_TASKS := gem5-clean qemu-clean ns-3-clean femu-clean
```
[the remark this answers is not in the document]
Comment (JLin-Li) [reply sims/external/rules.mk:119]: Thanks!

File: sims/external/rules.mk [hunk 2]
```diff
@@ -100,6 +102,19 @@ femu-clean:
 	-+cd $(EXTERNAL_SIMS_DIR)femu && $(MAKE) clean
 	rm -f $(EXTERNAL_SIMS_DIR)femu/ready
 
+$(d)bmv2:
+	git clone https://github.com/simbricks/bmv2.git $@
+
+$(d)bmv2/ready: $(d)bmv2 $(lib_netif)
+	+cd $< && ./autogen.sh && \
+	CPPFLAGS=-I$(abspath $(lib_dir)) ./configure && \
+	make -j
```
[the remark this answers is not in the document]
Comment (JLin-Li) [reply sims/external/rules.mk:111]: Sounds good!

File: docker/Dockerfile.buildenv [hunk 3]
```diff
@@ -53,6 +58,15 @@ RUN cd /tmp \
...
  && rm -rf /tmp/verilator
+RUN cd /tmp \
+ && git clone https://github.com/apache/thrift.git \
+ && cd thrift \
+ && ./bootstrap.sh \
+ && ./configure \
+ && make -j`nproc` \
+ && make install \
+ && rm -rf /tmp/thrift
+ ENV LD_LIBRARY_PATH=/usr/local/lib
```
[the remark this answers is not in the document]
Comment (JLin-Li) [reply docker/Dockerfile.buildenv]: Resolved!

**Review by jonas-kaufmann — APPROVED** [review]

**Pull request merged**